Commit for AED PDP, new AED knob and AED test fixes

fkeceli opened the pull request:
- Implemented the AED progressive dynamic-programming table optimization implemented in StringEditPeriodicityDetector class which speeds up the algorithm without changing its output. (Formerly known as online AED)
- Added a new knob in the StringEditEpochRecordFilter class: min_detectable_period.
- Some AED related tests were not included in the test/Makefile. Adding these tests to the Makefile exposed broken tests, fixed those tests.

Commit:
Fix broken ronn generation for out of source builds

Fixes #820

Change-Id: I55214db153a12308b7950fe5f8d6681a00501b44
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

dianarg committed:
Add script to plot different profiles against a baseline

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

Commit:
Fixup references to scripts/setup.py

Change-Id: I62d8420f2a5aab06a2581b2d49e25240c7b4d512
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

fkeceli committed:
Commit for AED PDP, new AED knob and fixing AED tests.

1. Implemented the AED progressive dynamic-programming table optimization implemented in StringEditPeriodicityDetector class which speeds up the algorithm without changing its output. (Formerly known as online AED)
2. Added a new knob in the StringEditEpochRecordFilter class: min_detectable_period.
3. Some AED related tests were not included in the test/Makefile. Adding these tests to the Makefile exposed broken tests, fixed those tests.

Commit:
Fix combination of bash conditionals

Change-Id: I5c807254b8b9ba84cff54e155b3a4fff81ad73ab
Signed-off-by: Brandon Baker <brandon.baker@intel.com>

cmcantalupo committed:
Change AppConf to pipe stdout and stderr to a file in bash template

Change-Id: I6a8ee704931671534b2668451bcd52b6d9d052c5
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Enable user specification of the number of iterations in the run_power_balancer_energy_nekbone.py script

Change-Id: I6c9a0024645bb0a1ad23f9693a49a895c63866fd
Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

fkeceli committed:
Converted the usage of new operator to make_unique for cleaner code

dianarg commented on src/EditDistEpochRecordFilter.cpp:
```diff
@@ -167,7 +194,7 @@ namespace geopm
         //
         if (!m_is_period_detected) {
-            if (m_edpd->get_score() >= m_edpd->get_period()) {
+            if (m_edpd->get_score() >= m_edpd->get_period() or m_edpd->get_period() < m_min_detectable_period) {
```
prefer `||` instead of `or`

fkeceli replied:
Changed in the last commit.

dianarg commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -45,70 +46,94 @@ namespace geopm
+        , m_record_count(0)
     {
-
+        m_DP = geopm::make_unique<std::vector<uint32_t> >(history_buffer_size * history_buffer_size * history_buffer_size);
```
Now that I see the context for this, I think this can be declared on the stack with
```std::vector<uint32_t> m_DP;``` in the header, and initialized in the list above with ```m_DP(history_buffer_size * history_buffer_size * history_buffer_size)```.  Then you don't need to dereference below

fkeceli committed:
Addressed comments for PR1264: redefined m_DP and changed or to |||

cmcantalupo committed:
Add automatic inference of reference target to energy efficiency plotting

- Commit removes data from the summary that was specific to the
  frequency map runs.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move power_balancer_energy scripts into energy_efficiency module

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add iteration option to all run scripts

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a warning and some error checking when inferring baseline profile name

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dianarg committed:
Change --show-profiles to --list-profiles

Signed-off-by: Diana Guttman <diana.r.guttman@intel.com>

fkeceli commented on src/EditDistEpochRecordFilter.cpp:
```diff
@@ -189,11 +216,11 @@ namespace geopm
                  (m_period_stable == m_stable_period_hysteresis * m_min_hysteresis_base_period)) ||
                 (m_edpd->get_period() > m_min_hysteresis_base_period &&
-                 (m_period_stable == m_stable_period_hysteresis * m_edpd->get_period()))) {
+                 (m_period_stable >= m_stable_period_hysteresis * m_edpd->get_period()))) {
```
Round m_stable_period_hysteresis * m_edpd->get_period()

fkeceli commented on src/EditDistPeriodicityDetector.hpp:
```diff
@@ -58,14 +58,24 @@ namespace geopm
+            int m_record_count;
+            std::vector<uint32_t> m_DP;
+
+            uint32_t m_myinf;
```
Replace this with UINT32_MAX in stdint.h.

fkeceli replied:
Replaced with numeric_limits<uint32_t>::max().

fkeceli commented on src/EditDistPeriodicityDetector.hpp:
```diff
@@ -58,14 +58,24 @@ namespace geopm
             int find_smallest_repeating_pattern(int index) const;
 
             CircularBuffer<uint64_t> m_history_buffer;
+            int m_history_buffer_size;
```
Make this a const.

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -37,6 +37,7 @@
 #include <cmath>
 #include <algorithm>
 
+#include "Helper.hpp"
```
Probably can remove this.

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -45,70 +46,95 @@ namespace geopm
     {
-
+        //m_DP = geopm::make_unique<std::vector<uint32_t> >(history_buffer_size * history_buffer_size * history_buffer_size);
+        m_myinf = 2*history_buffer_size;
```
Remove this.

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -45,70 +46,95 @@ namespace geopm
     {
         if (record.event == EVENT_REGION_ENTRY) {
             m_history_buffer.insert(record.signal);
+            m_record_count ++;
```
++m_record_count

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -45,70 +46,95 @@ namespace geopm
     }
 
+    void EditDistPeriodicityDetector::Dset(int ii, int jj, int mm, uint32_t val) {
+        m_DP[((ii % m_history_buffer_size) * m_history_buffer_size + (jj % m_history_buffer_size)) * m_history_buffer_size + (mm % m_history_buffer_size)] = val;
```
Add assertion to check boundaries. Get the name of the ASSERTION function from Diana.

Split up the line to make more readable.

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -45,70 +46,95 @@ namespace geopm
+        m_DP[((ii % m_history_buffer_size) * m_history_buffer_size + (jj % m_history_buffer_size)) * m_history_buffer_size + (mm % m_history_buffer_size)] = val;
+    }
+
+    uint32_t EditDistPeriodicityDetector::Dget(int ii, int jj, int mm) {
```
Add assertion to check boundaries.

fkeceli replied:
Not done.

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -45,70 +46,95 @@ namespace geopm
-                                               DD[ii - 1][jj - 1][mm] + term});
+        for (int mm = std::max({1, m_record_count-m_history_buffer_size}); mm < m_record_count; ++mm) {
+            for (int ii = std::max({1, m_record_count-m_history_buffer_size}); ii < mm+1; ++ii) {
+                int term;
```
initialize term to 2 and get rid of the ternary operator.

mm+1 --> mm + 1

fkeceli commented on src/EditDistPeriodicityDetector.cpp:
```diff
@@ -126,13 +152,20 @@ namespace geopm
         std::vector<uint64_t> recs = m_history_buffer.make_vector(
-            slice_start - 1, m_history_buffer.size());
+            slice_start, m_history_buffer.size());
+        std::vector<int> reps;
```
Delete this line -- used in record squashing.

fkeceli committed:
Addressed first round of comments for PR1264

fkeceli committed:
Fixed INF bug in PeriodicityDetector
